zerver/migrations/0417_backfill_message_realm.py

```diff
@@ -0,0 +1,75 @@
+from django.db import migrations, models, transaction
+from django.db.backends.postgresql.schema import BaseDatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+
+def backfill_message_realm(apps: StateApps, schema_editor: BaseDatabaseSchemaEditor) -> None:
+    RECIPIENT_HUDDLE = 3
+
+    Message = apps.get_model("zerver", "Message")
+    Realm = apps.get_model("zerver", "Realm")
+    Stream = apps.get_model("zerver", "Stream")
+    Subscription = apps.get_model("zerver", "Subscription")
+    UserProfile = apps.get_model("zerver", "UserProfile")
+    BATCH_SIZE = 10000
+
+    if not Message.objects.exists():
+        # Nothing to do if there are no messages yet.
+        return
+
+    def set_realm_batched(query: models.query.QuerySet[Message], realm_id: int) -> int:
+        count = 0
+        while True:
+            with transaction.atomic():
+                message_ids = set(
+                    query.filter(realm=None)
+                    .select_for_update()
+                    .values_list("id", flat=True)[:BATCH_SIZE]
+                )
+                if not message_ids:
+                    break
+                count += len(message_ids)
+                Message.objects.filter(id__in=message_ids).update(realm_id=realm_id)
+
+        return count
+
+    for realm_id in Realm.objects.values_list("id", flat=True):
+        print(f"\nBackfilling Message.realm for realm {realm_id}")
+        stream_recipient_ids = Stream.objects.filter(realm_id=realm_id).values_list(
+            "recipient_id", flat=True
+        )
+        query = Message.objects.filter(recipient_id__in=stream_recipient_ids)
+        stream_count = set_realm_batched(query, realm_id)
+
+        personal_recipient_ids = UserProfile.objects.filter(realm_id=realm_id).values_list(
+            "recipient_id", flat=True
+        )
+        query = Message.objects.filter(recipient_id__in=personal_recipient_ids)
+        personal_count = set_realm_batched(query, realm_id)
+
+        # Huddles don't have a direct way of finding their realm, so we have to
+        # go through the Subscription table.
+        huddle_recipient_ids = Subscription.objects.filter(
+            recipient__type=RECIPIENT_HUDDLE, user_profile__realm_id=realm_id
+        ).values_list("recipient_id", flat=True)
+        query = Message.objects.filter(recipient_id__in=huddle_recipient_ids)
+        huddle_count = set_realm_batched(query, realm_id)
+
+        print(
+            f"Updated {stream_count} stream messages, {personal_count} personal messages, and {huddle_count} huddle messages."
+        )
```

Comment:
So I think this algorithm is missing private messages that were sent by a user to a cross-realm bot like Welcome Bot; that specific class of message should instead be processed as being in the realm of the sender.

I think if we run a loop over just those messages first, before the `personal_recipient_ids` loop, we should be good. 

We also need to make sure the system bot realm is skipped in the loop here.

(I think huddles can't contain them, but even if they did, it should be OK, since only one non-system-bot realm will contain those messages)

Reply:
Hmm, I pushed a commit with appropriate adjustments, marked `squash or drop`. Some notes:
1. I kind of wish it wasn't necessary, because while it makes sense that a user's message to a cross-realm bot may be thought of as in the user's realm, it breaks the very nice invariant that `Message.realm = realm of Message.recipient` and introduces an easy special case to forget about and trip up on :thinking: 
2. We can't just exclude the system bot realm, because it is possible for `Messages` to exist within it, so it does have to be processed in the migration - because we have code snippets like this:
```
    # We also send a notification to the Zulip administrative realm
    admin_realm = get_realm(settings.SYSTEM_BOT_REALM)
    admin_realm_sender = get_system_bot(sender_email, admin_realm.id)
    try:
        # Check whether the stream exists
        signups_stream = get_signups_stream(admin_realm)
        # We intentionally use the same strings as above to avoid translation burden.
        with override_language(admin_realm.default_language):
            message = _("{user} just signed up for Zulip. (total: {user_count})").format(
                user=f"{user_profile.full_name} <`{user_profile.email}`>", user_count=user_count
            )
        internal_send_stream_message(
            admin_realm_sender, signups_stream, user_profile.realm.display_subdomain, message
        )
```
I resolved this by making sure the realm gets processed as the last one - that should make everything get set correctly in the sense of this approach.

Reply:
1. I believe the correct way to think about PMs to system bots is they belong in the realm of the user who sent the PM; that's the only version that makes sense in a world where we later move system bots into normal realms.

Regarding 2: I see; that code path only runs in the case that the system bot realm is also a normal realm ... but that does occur; OK. This strategy makes sense.